de/nmichael/efa/gui/EfaCloudSynchDialog.java

```diff
@@ -0,0 +1,206 @@
+            // Add download button.
+            if (isEfaCloud) {
+                guiItems.add(
+                        item = new ItemTypeButton(DOWNLOAD_BUTTON, IItemType.TYPE_PUBLIC, category,
+                                International.getString("Daten vom efaCloud Server laden")));
+                ((ItemTypeButton) item).setIcon(getIcon(IMAGE_DOWNLOAD));
+                item.setPadding(0, 0, 20, 20);
+                item.registerItemListener(this);
+                item.setFieldGrid(2, GridBagConstraints.SOUTH, GridBagConstraints.NONE);
+            }
+        }
+        super.setItems(guiItems);
+    }
+
+    /**
+     * The action triggered by the superclass on button pressed.
+     *
+     * @param itemType the item which triggered the event
+     * @param event    the event triggered. Only action events lead to execution of up-/download.
+     */
+    public void itemListenerAction(IItemType itemType, AWTEvent event) {
+
+        if (event instanceof ActionEvent) {
+            // authorization check
+            if (!admin.isAllowedAdministerProjectLogbook()) {
+                EfaMenuButton.insufficientRights(admin, itemType.getDescription());
+                return;
+            }
+            // Project record initialization check
+            if (Daten.project == null) return;
+            Project pr = Daten.project;
+
+            // if the current storage type is xml, this method is called to make it efacloud
+            // That will also trigger server uplaod
+            if (pr.getProjectStorageType() == IDataAccess.TYPE_FILE_XML) {
+                // read the entered parameters into the project record.
+                pr.setProjectEfaCloudURL(
+                        ((ItemTypeString) super.getItem("efaCloudURL")).getValue());
+                pr.setProjectStorageUsername(
+                        ((ItemTypeString) super.getItem("efaCloudUsername")).getValue());
+                pr.setProjectStoragePassword(
+                        ((ItemTypeString) super.getItem("efaCloudPassword")).getValue());
+                pr.setProjectStorageType(IDataAccess.TYPE_EFA_CLOUD);
+                // Store the changed settings
+                String prjName = pr.getName();
+                try {
+                    pr.close();
+                    Project.openProject(prjName, false);
+                    pr = Daten.project;
+                    if (parent instanceof EfaBaseFrame) {
+                        ((EfaBaseFrame) parent).openLogbook(pr.getCurrentLogbook());
+                    }
+                } catch (EfaException e) {
+                    de.nmichael.efa.util.Dialog.error(LogString.fileSavingFailed(prjName, International.getString("Projekt"), e.toString()));
+                    return;
+                }
+                // Start the server upload.
+                EfaCloudSynch.runEfaCloudSynchTask(this, EfaCloudSynch.Mode.upload, true, false, 0L,
```

Comment:
@tfyh This may need some improvement. In my testing, if often happened that the synchronization failed (for whatever reason, like me entering the wrong password or URL). While that will print an error, it doesn't undo the previous steps to update the project. I guess you have to first update the project before kicking off the synchronization (otherwise I'd just switch the order). So I guess we'd need to explicitly reset the storage type to what it used to be in case of fatal synchronization errors.

Reply:
Yes. I will simplify the synchronisation. It is too complex right now for a robust operation.